src/Compilers/CSharp/Portable/Emitter/Model/ArrayTypeSymbolAdapter.cs

```diff
@@ -11,14 +11,68 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.Symbols
 {
-    internal partial class ArrayTypeSymbol :
+    internal partial class
+#if DEBUG
+        ArrayTypeSymbolAdapter : SymbolAdapter,
+#else
+        ArrayTypeSymbol :
+#endif 
         Cci.IArrayTypeReference
+    {
+#if DEBUG
+        internal ArrayTypeSymbolAdapter(ArrayTypeSymbol underlyingArrayTypeSymbol)
+        {
+            AdaptedArrayTypeSymbol = underlyingArrayTypeSymbol;
+        }
+
+        internal sealed override Symbol AdaptedSymbol => AdaptedArrayTypeSymbol;
+        internal ArrayTypeSymbol AdaptedArrayTypeSymbol { get; }
+#else
+        internal ArrayTypeSymbol AdaptedArrayTypeSymbol => this;
+#endif 
+    }
+
+    internal partial class ArrayTypeSymbol
+    {
+#if DEBUG
+        private ArrayTypeSymbolAdapter? _lazyAdapter;
+
+        protected sealed override SymbolAdapter GetCciAdapterImpl() => GetCciAdapter();
+#endif
+
+        internal new
+#if DEBUG
+            ArrayTypeSymbolAdapter
+#else
+            ArrayTypeSymbol
+#endif
+            GetCciAdapter()
+        {
+#if DEBUG
+            if (_lazyAdapter is null)
+            {
+                return InterlockedOperations.Initialize(ref _lazyAdapter, new ArrayTypeSymbolAdapter(this));
+            }
+
+            return _lazyAdapter;
+#else
+            return this;
+#endif
+        }
```

Comment:
The `GetCciAdapter()` method definitions (here and in other SymbolAdapter classes) might be easier to read if the `#if DEBUG` and `#else` blocks were completely distinct. The `#if DEBUG` block could be merged with the `_lazyAdapter` block above.

Reply:
> The GetCciAdapter() method definitions (here and in other SymbolAdapter classes) might be easier to read if the #if DEBUG and #else blocks were completely distinct. The #if DEBUG block could be merged with the _lazyAdapter block above.

I am planning consolidating conditional compilation in a separate PR.


---
In reply to: [512876815](https://github.com/dotnet/roslyn/pull/48838#discussion_r512876815) [](ancestors = 512876815)